### PR TITLE
Add Flake

### PR DESCRIPTION
**Pull request by pnotequalnp**
Adds a flake to `/nix`. Contains outputs for the package itself as well as a devShell, overlay, and the existing NixOS module. This required modifying the NixOS module to use `pkgs.kmonad` as the default package instead of `import`ing `/nix/default.nix`, as that derivation is impure due to its pinning of Nixpkgs.

Also removes a deprecated use of `stdenv.lib` in `/nix/kmonad.nix` and fixes a typo in `/nix/nixos-module.nix`.

I'm not sure how people are using the NixOS module manually without a flake, and consequently don't know how this change would impact that usage. I would imagine that at worst, you would have to bring `kmonad` into `pkgs` via an overlay that imported `/nix/default.nix`.

### CODE REVIEW

**Comment by pnotequalnp** [comment]
Squashed to a single commit.

**Comment by flexagoon** [comment]
> I'm not sure how people are using the NixOS module manually without a flake

It's not hard at all. I just add

**Comment by flexagoon** [comment]
> I'm not sure how people are using the NixOS module manually without a flake

It's not hard at all. I just add
```
''${builtins.fetchGit "https://github.com/kmonad/kmonad"}/nix/nixos-module.nix''
```
to my `imports`

**Comment by pnotequalnp** [comment]
> > I'm not sure how people are using the NixOS module manually without a flake
> 
> It's not hard at all. I just add
> 
> ```
> ''${builtins.fetchGit "https://github.com/kmonad/kmonad"}/nix/nixos-module.nix''
> ```
> 
> to my `imports`

Interesting. So I believe this change would break that, right? For both to work, It would be simple just to rename the existing `nixos-module.nix` in this branch, and replace it with a module that simply brings in kmonad impurely via an overlay, a la the original module on master, and imports the renamed module. I'll push that change. If you'd be willing to test it, I'd appreciate that.

File: nix/module-base.nix [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            # prettify the service's name by taking the config filename...
+            conf-file = lists.last (strings.splitString "/" (toString kbd-path));
+            # ...and dropping the extension
+            conf-name = lists.head (strings.splitString "." conf-file);
+          in {
+          name = "kmonad-" +conf-name;
+          value = {
+            enable = true;
+            description = "KMonad Instance for: " +conf-name;
+            serviceConfig = {
```
Comment (flexagoon) [inline nix/module-base.nix:87]: This is missing my commit (https://github.com/kmonad/kmonad/pull/293)

**Comment by flexagoon** [comment]
Trying to build kmonad from your branch gives me the following error
```
error: attribute 'override' missing, at /nix/store/f3lpdi46776sg215lf8ybj3s24zs6a7p-nixos-21.11pre299299.3a8d7958a61/nixos/pkgs/top-level/all-packages.nix:33:16
(use '--show-trace' to show detailed location information)
building Nix...
error: attribute 'isStatic' missing, at /nix/store/f3lpdi46776sg215lf8ybj3s24zs6a7p-nixos-21.11pre299299.3a8d7958a61/nixos/pkgs/tools/package-management/nix/default.nix:26:20
(use '--show-trace' to show detailed location information)
building the system configuration...
error: undefined variable 'isStatic' at /nix/store/f3lpdi46776sg215lf8ybj3s24zs6a7p-nixos-21.11pre299299.3a8d7958a61/nixos/pkgs/tools/networking/curl/default.nix:13:6
(use '--show-trace' to show detailed location information)
```

**Comment by Bobbbay** [comment]
+1 still relevant.

If this project was flaked that would be great, it would enable ~~elitists~~ flake users to add KMonad in a pinch.

Anyways, @pnotequalnp, how are you running this? I'd like to be able to say `nix run github:pnotequalnp/flake`, but since `flake.nix` is kept in `/nix` Nix doesn't see it. I may be missing an option to specify the location of the flake, but a skim through nix run's docs shows nothing relevant. 

@flexagoon, are you evaluating impurely? When I try it with the current NixOS module, I get:

```
error: in pure evaluation mode, 'fetchTree' requires an immutable input, at /nix/store/6px4xrrq4krngx6k0qw1aqaf3q0kgxbs-source/host/NotYourLaptop/default.nix:4:48
(use '--show-trace' to show detailed location information)
error: getting status of '/tmp/nixos-rebuild.SVnnJF/result': No such file or directory
```

(while, of course, this evaluates purely.)

danke for your work!

**Comment by pnotequalnp** [comment]
I have no idea what could be causing the error @flexagoon and don't have an impure config readily available to test it.

> Anyways, @pnotequalnp, how are you running this? I'd like to be able to say `nix run github:pnotequalnp/flake`, but since `flake.nix` is kept in `/nix` Nix doesn't see it. I may be missing an option to specify the location of the flake, but a skim through nix run's docs shows nothing relevant.

You need [`?dir=nix`](https://github.com/pnotequalnp/nix-config/blob/main/flake.nix#L27).

> When I try it with the current NixOS module

What are you doing to get this error?

**Comment by Bobbbay** [comment]
> You need ?dir=nix.

That was it, thanks!

> What are you doing to get this error?

1. Add `''${builtins.fetchGit "https://github.com/kmonad/kmonad"}/nix/nixos-module.nix''` to my imports
2. Enable `services.kmonad`
3. Rebuild
4. ~~Profit~~ Error :(

**Comment by pnotequalnp** [comment]
> > You need ?dir=nix.
> 
> That was it, thanks!
> 
> > What are you doing to get this error?
> 
>     1. Add `''${builtins.fetchGit "https://github.com/kmonad/kmonad"}/nix/nixos-module.nix''` to my imports
> 
>     2. Enable `services.kmonad`
> 
>     3. Rebuild
> 
>     4. ~Profit~ Error :(

Yes I'd expect an error if you import like that and your config is in a flake. The original derivation isn't pure which is why I had to pin Nixpkgs via the flake and change the NixOS module to use that instead.

**Comment by dlip** [comment]
I'm adding kmonad to my flake in my own hacky way until this is in https://www.github.com/dlip/nixconfig/tree/master/flake.nix

**Comment by pnotequalnp** [comment]
> I'm adding kmonad to my flake in my own hacky way until this is in https://www.github.com/dlip/nixconfig/tree/master/flake.nix

If you don't mind being some commits behind master you can also [use this branch directly](https://github.com/pnotequalnp/nix-config/blob/main/flake.nix#L27).

**Comment by slotThe** [comment]
@pnotequalnp what's the status of this?  As far as I can see, @flexagoon's and @Bobbbay's problems have been solved?  If yes, could you rebase on top of current master so that we can merge?

**Comment by pnotequalnp** [comment]
> @pnotequalnp what's the status of this? As far as I can see, @flexagoon's and @Bobbbay's problems have been solved? If yes, could you rebase on top of current master so that we can merge?

Rebased onto master and improved the flake. I successfully tested:
- [x] Building via the flake on x86_64-linux and aarch64-linux
- [x] Building via `default.nix` on x86_64-linux
- [x] The NixOS module on x86_64-linux

If someone wants to test something else before merging (mainly the `fetchGit` method of importing the NixOS module into non-flake NixOS configs) that might be a good idea.

**Comment by slotThe** [comment]
Sorry for not getting to this until now—looks great, thanks!

**Comment by dlip** [comment]
I just tested this but getting an error since kmonad.nix doesn't exist

```
let kmonad = import ./kmonad.nix;
```

**Comment by pnotequalnp** [comment]
> I just tested this but getting an error since kmonad.nix doesn't exist
> 
> ```
> let kmonad = import ./kmonad.nix;
> ```

Indeed f49a41c was added since testing, completely redoing the Nix build files, and is incompatible with this flake. I'm not familiar with the new way the `default.nix` is structured off the top of my head to get them to mesh.

**Comment by dlip** [comment]
I've created a fix pr #364

**Pull request merged**